### ISSUE TITLE
Uninitialized bytes get copied to the buffer when creating an uninitialized tuple

### ISSUE THREAD

**Issue by MSxDOS** [rust-lang/rust#83087]
In this example:
```Rust
use std::mem::MaybeUninit;

extern {
    fn fill(p: *mut u16);
}

pub unsafe fn test() {
    let (mut buffer, mut buffer2): ([u16; 512], [u16; 256]) = MaybeUninit::uninit().assume_init();
    fill(buffer.as_mut_ptr());
    fill(buffer2.as_mut_ptr())
}
```
there is a call to `memcpy` for one of the buffers that copies uninitialized data to it.
https://rust.godbolt.org/z/z1zcor

**Comment by MSxDOS** [comment]
cc @nikic  

This started after https://github.com/rust-lang/rust/pull/82806 was merged.

**Comment by nikic** [comment]
Upstream fix: https://github.com/llvm/llvm-project/commit/55566609711f284581d822576705354444e6a7f0

**Comment by pietroalbini** [comment]
This slipped up to stable 1.52.0, apparently we didn't label it `regression-from-stable-to-beta` losing track of it :/

**Comment by nikic** [comment]
@pietroalbini We don't plan to address this prior to the next LLVM upgrade anyway. It's just an optimization issue.

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2383087.20Uninitialized.20bytes.20get.20copied.20to.20the.20buffer.20when.20cr.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-high +T-compiler

**Comment by nikic** [comment]
Fixed by #87570, new IR does not have memcpy:

```
define void @_ZN5test24test17h0962739e00df972bE() unnamed_addr #0 {
start:
  %buffer2 = alloca [256 x i16], align 2
  %buffer = alloca [512 x i16], align 2
  %0 = bitcast [512 x i16]* %buffer to i8*
  call void @llvm.lifetime.start.p0i8(i64 1024, i8* nonnull %0)
  %1 = bitcast [256 x i16]* %buffer2 to i8*
  call void @llvm.lifetime.start.p0i8(i64 512, i8* nonnull %1)
  %2 = getelementptr inbounds [512 x i16], [512 x i16]* %buffer, i64 0, i64 0
  call void @fill(i16* nonnull %2)
  %3 = getelementptr inbounds [256 x i16], [256 x i16]* %buffer2, i64 0, i64 0
  call void @fill(i16* nonnull %3)
  call void @llvm.lifetime.end.p0i8(i64 512, i8* nonnull %1)
  call void @llvm.lifetime.end.p0i8(i64 1024, i8* nonnull %0)
  ret void
}
```
